### DISCUSSION TITLE
Make a script to automate the process?

### DISCUSSION THREAD

**Discussion by cmspam** [fakemanhk/openwrt-jp-ipoe#9] [Ideas]
It might be a good idea to write a script to automate the process.

It would do something along the lines of:

0. Check that the version of openwrt is recent.
1. opkg update && opkg install map curl
2. Get the ipv6 address from the wan6 interface
3. Use curl to query http://ipv4.web.fc2.com/map-e.html with the ipv6 address to get the necessary parameters.
4. Fill the parameters into /etc/config/network to create the map-wan interface.
5. Fetch the firewall.user and map.sh files, place them in the correct locations.
6. Update /etc/config/firewall to use the firewall.user file
7. Reboot

Then perhaps this could be distributed as a package for openwrt, or in some other way so that beginners who don't want to ssh into the router can use it.

It should not be too hard, but I might not have time to write and test it. If anyone wants to... ;)

**Comment by cmspam** [comment]
Looking into it a little bit more,  http://ipv4.web.fc2.com/map-e.html  is using javascript code to calculate all of the parameters. So the javascript code would need to be converted into a bash script to run locally, or it would need node (which is 11MB on routers with potentially very small storage) as a dependency and just run the javascript code to do the calculations. Or a webserver could be set up to respond to a query of the ipv6 address with the right information...

It's too much hassle for me, and there are various potential points of failure, so although I thought about it, I'm probably not going to make an automation script. I guess we can focus on making it as easy as possible with clear instructions.

**Reply by strxno** [reply]
I wonder if there's a dump of some japanese router firmware which might have that script already made for us.

**Reply by cmspam** [reply]
That's an interesting idea.
Really, the necessary information is all available in the javascript code on the linked site. It would just be quite a bit of a pain to translate that to a shell script that openwrt could run without any addons... But I suspect with other scripts from other routers, it would also need to be translated and modified to run.
Maybe I will work on it if I have time though, just for fun.

**Reply by strxno** [reply]
IO-Data WN-DX1200GR runs OpenWRT copy, so it should be similar. I have just extracted the firmware and am looking into it.

**Reply by cmspam** [reply]
Whatever the function is should have a list of what the applicable ip addresses will start with: start with 240b, or start with 2404, or start with 2400.

**Reply by cmspam** [reply]
I checked the same firmware, but it's in the binary file /usr/sbin/map-e-config I believe. Unfortunately, it's a binary.

**Reply by strxno** [reply]
Could it be /usr/bin/mapcalc on OpenWRT? We already have that file?

**Reply by cmspam** [reply]
It's probably not in OpenWRT, because there is data needed specific to the various Japanese ISPs.
I'm re-writing the javascript code to python, because a python environment can be established in openwrt with low memory. I'll write python script to automate everything.

If you are using an ISP which is OCN or BIGLOBE based, I hope you can test the calculator and make sure the results are right.

**Reply by strxno** [reply]
[4o6.sh.txt](https://github.com/fakemanhk/openwrt-jp-ipoe/files/10813116/4o6.sh.txt)

What about this file? Looks more like it...

**Reply by cmspam** [reply]
Yes, that file looks correct. It seems it depends on some other files maybe and API keys, and it requests to api.enabler.ne.jp for V6Plus for example
Very interesting.
EDIT: The API request seems to be for people with static IPs as far as I can tell. It's still interesting to look into though.

**Reply by strxno** [reply]
Using the code from url in line 2583 gave a nice output in JSON of prefixes as well :)

**Reply by strxno** [reply]
Interesting... It doesn't match what the online calculator gave me.

**Reply by cmspam** [reply]
> Using the code from url in line 2583 gave a nice output in JSON of prefixes as well :)

2583 will work with OCN, it looks like we need this for v6plus query.

v6PLUS_UNIQUE_ID="${WORKDIR}/v6plus_unique_id"

Do you find this file?

**Reply by strxno** [reply]
https://rule.map.ocn.ad.jp/?ipv6Prefix=2400:4053:xxxx:xxxx::&ipv6PrefixLength=64&code=codefromline2583 outputs me the map rules, just all of them for all prefixes. Seems that v6plus_unique_id gets generated by the script?

**Reply by cmspam** [reply]
Unfortunately I don't have a similar way to do this for JPNE.
Where did you find this sh file? I checked an unpacked firmware image but didn't find it. Can you let me know the image you got it from?

**Reply by strxno** [reply]
https://dl.elecom.co.jp/support/download/network/wireless_lan/wrc-2533gst2/WRC-2533GST2_v1.30.zip

**Comment by cmspam** [comment]
I have made a python version to calculate the parameters. It is based off of the code written in javascript att http://ipv4.web.fc2.com/map-e.html

I can use this to make automated setup for map-e configuration, but can some people please test that you are getting the correct results, the same as the above website, when trying with your own IP? I can confirm it works for mine, but mine is a 240b (JPNE), and I need some people with other addresses to also try.

Just rename to calculate.py and run with python3
It will ask you to input ipv6 address, and will give you same results as the website, which are needed for openwrt config.


[calculate.py.txt](https://github.com/fakemanhk/openwrt-jp-ipoe/files/10813364/calculate.py.txt)

**Reply by strxno** [reply]
Yes it calculted mine correctly. Mine starts with 2400:4053

**Reply by cmspam** [reply]
Thanks for confirming. Then I have to think, whether to make a script to run on the computer (which connects to openwrt over ssh and does it) or one to run on openwrt itself. Maybe the former is easier for most people.

**Reply by fakemanhk** [reply]
My IP starts with 2400:4051,  the calculations also worked for me.

**Reply by TimOfAllTrades** [reply]
I will be on a business trip for the next few days, but I will also give this a try when I get back.

**Reply by fakemanhk** [reply]
Coming back to this thread, since not all routers having enough space for python packages, I am thinking if we can turn this python script to web app so that I can host in this GitHub (because we never know how long the original online calculator will be online).

**Reply by LIII-XXII** [reply]
This shouldn't need python or a webapp IMO.

There's a bash version here, though it looks hardcoded.

https://github.com/site-u2023/config-software/blob/main/map-e.sh

**Reply by fakemanhk** [reply]
Oh yeah, I looked into the author's page, the author was actually creating the script with reference to the guide here so it should be the same thing. Hardcoding isn't really bad thing if: 1) they don't change; 2) we are able to list all possibilities.
